apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/grpc/v1/server/TracingServerCall.java

```diff
@@ -0,0 +1,111 @@
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+import org.apache.skywalking.apm.plugin.grpc.v1.OperationNameFormatUtil;
+
+import static org.apache.skywalking.apm.plugin.grpc.v1.Constants.*;
+
+/**
+ * @author wang zheng, kanro
+ */
+public class TracingServerCall<REQUEST, RESPONSE> extends ForwardingServerCall.SimpleForwardingServerCall<REQUEST, RESPONSE> {
+
+    private final String operationPrefix;
+    private final ContextSnapshot contextSnapshot;
+
+    protected TracingServerCall(ServerCall<REQUEST, RESPONSE> delegate, ContextSnapshot contextSnapshot) {
+        super(delegate);
+        this.operationPrefix = OperationNameFormatUtil.formatOperationName(delegate.getMethodDescriptor()) + SERVER;
+        this.contextSnapshot = contextSnapshot;
+    }
+
+    @Override
+    public void sendMessage(RESPONSE message) {
+        // We just create the request on message span for server stream calls.
+        if (!getMethodDescriptor().getType().serverSendsOneMessage()) {
+            try {
+                final AbstractSpan span = ContextManager.createLocalSpan(operationPrefix + RESPONSE_ON_MESSAGE_OPERATION_NAME);
+                span.setComponent(ComponentsDefine.GRPC);
+                span.setLayer(SpanLayer.RPC_FRAMEWORK);
+                ContextManager.continued(contextSnapshot);
+                super.sendMessage(message);
+            } catch (Throwable t) {
+                ContextManager.activeSpan().errorOccurred().log(t);
+                throw t;
+            } finally {
+                ContextManager.stopSpan();
+            }
+        } else {
+            super.sendMessage(message);
+        }
+    }
+
+    @Override
+    public void close(Status status, Metadata trailers) {
+        try {
+            final AbstractSpan span = ContextManager.createLocalSpan(operationPrefix + RESPONSE_ON_CLOSE_OPERATION_NAME);
+            span.setComponent(ComponentsDefine.GRPC);
+            span.setLayer(SpanLayer.RPC_FRAMEWORK);
+            ContextManager.continued(contextSnapshot);
+
+            switch (status.getCode()) {
+                case OK:
+                    break;
+                // UNKNOWN/INTERNAL status code will case error in this span.
+                // Those status code means some unexpected error occurred in server.
+                // Similar to 5XX in HTTP status.
+                case UNKNOWN:
+                case INTERNAL:
+                    if (status.getCause() == null) {
+                        span.errorOccurred().log(status.asRuntimeException());
+                    } else {
+                        span.errorOccurred().log(status.getCause());
+                    }
+                    break;
+                // Other status code means some predictable error occurred in server.
+                // Like PERMISSION_DENIED or UNAUTHENTICATED somethings.
+                // Similar to 4XX in HTTP status.
+                default:
```

Comment:
`case UNKNOWN`, `case INTERNAL` and `default` seem to have the same logic. What is the difference?

Reply:
`UNKNOWN` and `INTERNAL` is similar to the `5xx` errors in HTTP, it must case span error even status has no cause exception.

`default`(others) is similar to the `4xx` errors in HTTP, we just case error when status has cause exception for user debugging.